staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -359,10 +361,33 @@ func newInformer(
 				switch d.Type {
 				case Sync, Added, Updated:
 					if old, exists, err := clientState.Get(d.Object); err == nil && exists {
-						if err := clientState.Update(d.Object); err != nil {
-							return err
+						oldMeta, err := meta.Accessor(old)
+						if err != nil {
+							return fmt.Errorf("old object has no meta: %v", err)
+						}
+						newMeta, err := meta.Accessor(d.Object)
+						if err != nil {
+							return fmt.Errorf("new object has no meta: %v", err)
+						}
+
+						if oldMeta.GetUID() == newMeta.GetUID() {
+							// This is the same object, send update event
+							if err := clientState.Update(d.Object); err != nil {
+								return err
+							}
+							h.OnUpdate(old, d.Object)
+						} else {
+							// If the object was recreated we need to first generate delete event followed by add event
+							if err := clientState.Delete(old); err != nil {
```

Comment:
wouldn't this be a DeletedFinalStateUnknown?

Reply:
what else is using/observing clientState concurrently? are there races that removing and then re-adding exposes us to?

Reply:
> wouldn't this be a DeletedFinalStateUnknown, since we didn't actually observe the delete?
yes, good catch

> what else is using/observing clientState concurrently? are there races that removing and then re-adding exposes us to?

they have the same key, same key is handles serially IIRC